repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/RequestHandler.scala

```diff
@@ -0,0 +1,524 @@
+              datastore.put(ADMIN_TOKEN_KEY, s.get)
+              s
+            case Failure(f) => Failure(IdentityCommuncationException("Token not found in identity response during Admin Authentication", f))
+          }
+        case Failure(x) => Failure(IdentityCommuncationException("Failure communicating with identity during Admin Authentication", x))
+      }
+    }
+  }
+
+  def extractTenant(requestUri: String): Option[String] = {
+    val maybeTenant = for {
+      tenantHandling <- Option(config.getTenantHandling)
+      validateTenant <- Option(tenantHandling.getValidateTenant)
+      uriExtractionRegex <- Option(validateTenant.getUriExtractionRegex)
+    } yield {
+        val regex = uriExtractionRegex.r
+        requestUri match {
+          case regex(tenantId, _*) => Option(tenantId)
+          case _ => Option.empty[String]
+        }
+      }
+    maybeTenant.flatten
+  }
+
+  def tenantAuthorization(expectedTenant: Option[String], validToken: ValidToken): Option[Try[Vector[String]]] = {
+    def buildTenantVector(defaultTenant: String, roleTenants: Seq[String], uriTenant: Option[String]): Vector[String] = {
+      val sendAllTenants = config.getTenantHandling.isSendAllTenantIds
+      val sendTenantIdQuality = Option(config.getTenantHandling.getSendTenantIdQuality)
+      val sendQuality = sendTenantIdQuality.isDefined
+      val defaultTenantQuality = sendTenantIdQuality.map(_.getDefaultTenantQuality).getOrElse(0.0)
+      val uriTenantQuality = sendTenantIdQuality.map(_.getUriTenantQuality).getOrElse(0.0)
+      val rolesTenantQuality = sendTenantIdQuality.map(_.getRolesTenantQuality).getOrElse(0.0)
+
+      var preferredTenant = defaultTenant
+      var preferredTenantQuality = defaultTenantQuality
+      uriTenant foreach { tenant =>
+        preferredTenant = tenant
+
+        preferredTenantQuality = if (defaultTenant.equals(tenant)) {
+          math.max(defaultTenantQuality, uriTenantQuality)
+        } else {
+          uriTenantQuality
+        }
+      }
+
+      if (sendAllTenants && sendQuality) {
+        val priorityTenants = uriTenant match {
+          case Some(tenant) => Vector(s"$defaultTenant;q=$defaultTenantQuality", s"$tenant;q=$uriTenantQuality")
+          case None => Vector(s"$defaultTenant;q=$defaultTenantQuality")
+        }
+        priorityTenants ++ roleTenants.map(tid => s"$tid;q=$rolesTenantQuality")
+      } else if (sendAllTenants && !sendQuality) {
+        Vector(defaultTenant) ++ roleTenants
+      } else if (!sendAllTenants && sendQuality) {
+        Vector(s"$preferredTenant;q=$preferredTenantQuality")
+      } else {
+        Vector(preferredTenant)
+      }
+    }
+
+    Option(config.getTenantHandling) map { tenantHandling =>
+      Option(tenantHandling.getValidateTenant) map { validateTenant =>
+        Option(validateTenant.getBypassValidationRoles) map {
+          _.getRole.asScala.intersect(validToken.roles).nonEmpty
+        } filter {
+          identity
```

Comment:
So the Option and map above yield an Option[boolean]. That means we could have the following results:
Some(true)
Some(false)
None

Well, we can execute the same logic for Some(false) as None, since both cases imply that a bypass role was not present. So by filtering using the identity function, we reduce the possible results to:
Some(true)
None

I thought it cleaned up the match a little bit, but it would be equivalent to make the match against Some(true) and _ instead.